### ISSUE TITLE
compilation failure while setting up security subsystem

### ISSUE THREAD

**Issue by alaniwi** [ESGF/esgf-installer#273]
_Listing here because I am not convinced that the issues on the esgf-security repo are actively monitored._

During an installation of v2.5.14-master-release including the idp component, the build failed during attempt to build psycopg2 in the conda environment. 

The rest of the ESGF installation process continued, but as a result of this failure it did not set up the `esgf_security.*` in postgres, so the IdP was not properly installed even if psycopg2 was somehow installed later.

```
Writing /tmp/easy_install-K0N8hL/psycopg2-2.7.3.1/setup.cfg
Running psycopg2-2.7.3.1/setup.py -q bdist_egg --dist-dir /tmp/easy_install-K0N8hL/psycopg2-2.7.3.1/egg-dist-tmp-9zwh_I
psycopg/psycopgmodule.c: In function 'psyco_libpq_version':
psycopg/psycopgmodule.c:399: warning: implicit declaration of function 'PQlibVersion'
psycopg/psycopgmodule.c: In function 'init_psycopg':
psycopg/psycopgmodule.c:891: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:894: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:897: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:900: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:903: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:906: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:909: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:912: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:915: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:918: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:921: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:924: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:927: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:930: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:933: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:936: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:939: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:942: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:945: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:949: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:952: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:989: warning: dereferencing type-punned pointer will break strict-aliasing rules
psycopg/psycopgmodule.c:891: warning: dereferencing pointer 'connectionType.84' does break strict-aliasing rules
psycopg/psycopgmodule.c:891: note: initialized from here
psycopg/psycopgmodule.c:894: warning: dereferencing pointer 'cursorType.85' does break strict-aliasing rules
psycopg/psycopgmodule.c:894: note: initialized from here
psycopg/psycopgmodule.c:897: warning: dereferencing pointer 'replicationConnectionType.86' does break strict-aliasing rules
psycopg/psycopgmodule.c:897: note: initialized from here
psycopg/psycopgmodule.c:900: warning: dereferencing pointer 'replicationCursorType.87' does break strict-aliasing rules
psycopg/psycopgmodule.c:900: note: initialized from here
psycopg/psycopgmodule.c:903: warning: dereferencing pointer 'replicationMessageType.88' does break strict-aliasing rules
psycopg/psycopgmodule.c:903: note: initialized from here
psycopg/psycopgmodule.c:906: warning: dereferencing pointer 'typecastType.89' does break strict-aliasing rules
psycopg/psycopgmodule.c:906: note: initialized from here
psycopg/psycopgmodule.c:909: warning: dereferencing pointer 'qstringType.90' does break strict-aliasing rules
psycopg/psycopgmodule.c:909: note: initialized from here
psycopg/psycopgmodule.c:912: warning: dereferencing pointer 'binaryType.91' does break strict-aliasing rules
psycopg/psycopgmodule.c:912: note: initialized from here
psycopg/psycopgmodule.c:915: warning: dereferencing pointer 'isqlquoteType.92' does break strict-aliasing rules
psycopg/psycopgmodule.c:915: note: initialized from here
psycopg/psycopgmodule.c:918: warning: dereferencing pointer 'pbooleanType.93' does break strict-aliasing rules
psycopg/psycopgmodule.c:918: note: initialized from here
psycopg/psycopgmodule.c:921: warning: dereferencing pointer 'pintType.94' does break strict-aliasing rules
psycopg/psycopgmodule.c:921: note: initialized from here
psycopg/psycopgmodule.c:924: warning: dereferencing pointer 'pfloatType.95' does break strict-aliasing rules
psycopg/psycopgmodule.c:924: note: initialized from here
psycopg/psycopgmodule.c:927: warning: dereferencing pointer 'pdecimalType.96' does break strict-aliasing rules
psycopg/psycopgmodule.c:927: note: initialized from here
psycopg/psycopgmodule.c:930: warning: dereferencing pointer 'asisType.97' does break strict-aliasing rules
psycopg/psycopgmodule.c:930: note: initialized from here
psycopg/psycopgmodule.c:933: warning: dereferencing pointer 'listType.98' does break strict-aliasing rules
psycopg/psycopgmodule.c:933: note: initialized from here
psycopg/psycopgmodule.c:936: warning: dereferencing pointer 'chunkType.99' does break strict-aliasing rules
psycopg/psycopgmodule.c:936: note: initialized from here
psycopg/psycopgmodule.c:939: warning: dereferencing pointer 'notifyType.100' does break strict-aliasing rules
psycopg/psycopgmodule.c:939: note: initialized from here
psycopg/psycopgmodule.c:942: warning: dereferencing pointer 'xidType.101' does break strict-aliasing rules
psycopg/psycopgmodule.c:942: note: initialized from here
psycopg/psycopgmodule.c:945: warning: dereferencing pointer 'errorType.102' does break strict-aliasing rules
psycopg/psycopgmodule.c:945: note: initialized from here
psycopg/psycopgmodule.c:949: warning: dereferencing pointer 'diagnosticsType.105' does break strict-aliasing rules
psycopg/psycopgmodule.c:949: note: initialized from here
psycopg/psycopgmodule.c:952: warning: dereferencing pointer 'lobjectType.106' does break strict-aliasing rules
psycopg/psycopgmodule.c:952: note: initialized from here
psycopg/psycopgmodule.c:989: warning: dereferencing pointer 'pydatetimeType.111' does break strict-aliasing rules
psycopg/psycopgmodule.c:989: note: initialized from here
psycopg/pqpath.c: In function 'pq_get_last_result':
psycopg/pqpath.c:1127: error: 'PGRES_COPY_BOTH' undeclared (first use in this function)
psycopg/pqpath.c:1127: error: (Each undeclared identifier is reported only once
psycopg/pqpath.c:1127: error: for each function it appears in.)
psycopg/pqpath.c: In function 'pq_fetch':
psycopg/pqpath.c:1931: error: 'PGRES_COPY_BOTH' undeclared (first use in this function)
error: Setup script exited with error: command 'gcc' failed with exit status 1
ERROR: Could not create esgf security python module

Sorry...
This action did not complete successfully
Please re-run this task until successful before continuing further

Also please review the installation FAQ it may assist you
https://github.com/ESGF/esgf.github.io/wiki/ESGFNode%7CFAQ
```

**Comment by alaniwi** [comment]
Just to be clear, this happens at the statement 

`$cdat_home/bin/easy_install ${esgf_security_egg_file}`

in function `configure_postgress` in `/usr/local/bin/esg-security`

Where I write above that `esgf_security.*` was not set up because of this, this is referring to the call to `esgf_security_initialize` later in the `configure_postgress` function, which was not reached.

**Comment by alaniwi** [comment]
In the latest postresql library (`git clone git://git.postgresql.org/git/postgresql.git`), I see:

in `src/interfaces/libpq/libpq-fe.h`:

```
typedef enum
{
        PGRES_EMPTY_QUERY = 0,          /* empty query string was executed */
        PGRES_COMMAND_OK,                       /* a query command that doesn't return
                                                                 * anything was executed properly by the
                                                                 * backend */
        PGRES_TUPLES_OK,                        /* a query command that returns tuples was
                                                                 * executed properly by the backend, PGresult
                                                                 * contains the result tuples */
        PGRES_COPY_OUT,                         /* Copy Out data transfer in progress */
        PGRES_COPY_IN,                          /* Copy In data transfer in progress */
        PGRES_BAD_RESPONSE,                     /* an unexpected response was recv'd from the
                                                                 * backend */
        PGRES_NONFATAL_ERROR,           /* notice or warning message */
        PGRES_FATAL_ERROR,                      /* query failed */
        PGRES_COPY_BOTH,                        /* Copy In/Out data transfer in progress */
        PGRES_SINGLE_TUPLE                      /* single tuple from larger resultset */
} ExecStatusType;
```

but on the data node, it appears that it is using the `/usr/include/libpq-fe.h` from the RPM, which is `postgresql-devel-8.4.20-7.el6.x86_64`, and this has:

```
typedef enum
{
        PGRES_EMPTY_QUERY = 0,          /* empty query string was executed */
        PGRES_COMMAND_OK,                       /* a query command that doesn't return
                                                                 * anything was executed properly by the
                                                                 * backend */
        PGRES_TUPLES_OK,                        /* a query command that returns tuples was
                                                                 * executed properly by the backend, PGresult
                                                                 * contains the result tuples */
        PGRES_COPY_OUT,                         /* Copy Out data transfer in progress */
        PGRES_COPY_IN,                          /* Copy In data transfer in progress */
        PGRES_BAD_RESPONSE,                     /* an unexpected response was recv'd from the
                                                                 * backend */
        PGRES_NONFATAL_ERROR,           /* notice or warning message */
        PGRES_FATAL_ERROR                       /* query failed */
} ExecStatusType;
```

**Comment by alaniwi** [comment]
Looking at various tags in the postgres repo, it seems that `PGRES_COPY_BOTH` was introduced in `9.1.0` (absent in `9.0.23`). No sign of a later `postgresql-devel` RPM, so the later library will have to be provided in the conda environment.

**Comment by alaniwi** [comment]
With strace, I can see, if I try to do the `gcc` command on which it falls over:

```
strace -f -e open -o strace.out gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -DPSYCOPG_DEFAULT_PYDATETIME=1 -DPSYCOPG_VERSION="2.7.3.1 (dt dec pq3 ext lo64)" -DPG_VERSION_NUM=90603 -DHAVE_LO64=1 -I/usr/local/conda/envs/esgf-pub/include/python2.7 -I. -I/usr/local/conda/envs/esgf-pub/include -I/usr/local/conda/envs/esgf-pub/include/server -c psycopg/pqpath.c -o build/temp.linux-x86_64-2.7/psycopg/pqpath.o -Wdeclaration-after-statement
```

it is using:
```
30301 open("/usr/include/libpq-fe.h", O_RDONLY|O_NOCTTY) = 4
```

If I do a `conda install postgresql`, it offers to install 9.5.4-0, which I say yes to. Repeating the above, I now get:
```
31450 open("/usr/local/conda/envs/esgf-pub/include/libpq-fe.h", O_RDONLY|O_NOCTTY) = 4
```
and the build succeeds.

**Comment by alaniwi** [comment]
I'm going to try an installation with a sleep inserted in `setup_subsystem` between downloading the subsystem script and running it, so that I can make a hand-edit to the `esg-security` script in order to insert the `conda install postgresql` line. I'll report here how I get on...

**Comment by alaniwi** [comment]
Well, that got past the psycopg2 build, but then failed just after:
```
Finished processing dependencies for esgf-security==0.1.5
```
with:
```
Traceback (most recent call last):
  File "/usr/local/conda/bin/esgf_security_initialize", line 4, in <module>
    __import__('pkg_resources').run_script('esgf-security==0.1.5', 'esgf_security_initialize')
  File "/usr/local/conda/lib/python2.7/site-packages/pkg_resources/__init__.py", line 748, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/usr/local/conda/lib/python2.7/site-packages/pkg_resources/__init__.py", line 1517, in run_script
    exec(code, namespace, namespace)
  File "/usr/local/conda/lib/python2.7/site-packages/esgf_security-0.1.5-py2.7.egg/EGG-INFO/scripts/esgf_security_initialize", line 100, in <module>
    main(sys.argv[1:])
  File "/usr/local/conda/lib/python2.7/site-packages/esgf_security-0.1.5-py2.7.egg/EGG-INFO/scripts/esgf_security_initialize", line 83, in main
    version_control(dburl, repo_path)
  File "<decorator-gen-8>", line 2, in version_control
  File "/usr/local/conda/lib/python2.7/site-packages/sqlalchemy_migrate-0.11.0-py2.7.egg/migrate/versioning/util/__init__.py", line 163, in with_engine
    engine = construct_engine(url, **kw)
  File "/usr/local/conda/lib/python2.7/site-packages/sqlalchemy_migrate-0.11.0-py2.7.egg/migrate/versioning/util/__init__.py", line 148, in construct_engine
    return create_engine(engine, **kwargs)
  File "build/bdist.linux-x86_64/egg/sqlalchemy/engine/__init__.py", line 338, in create_engine
  File "build/bdist.linux-x86_64/egg/sqlalchemy/engine/strategies.py", line 64, in create
  File "build/bdist.linux-x86_64/egg/sqlalchemy/dialects/postgresql/psycopg2.py", line 312, in dbapi
  File "/usr/local/conda/lib/python2.7/site-packages/psycopg2-2.7.3.1-py2.7-linux-x86_64.egg/psycopg2/__init__.py", line 50, in <module>
    from psycopg2._psycopg import (                     # noqa
ImportError: /usr/local/conda/lib/python2.7/site-packages/psycopg2-2.7.3.1-py2.7-linux-x86_64.egg/psycopg2/_psycopg.so: undefined symbol: PQconninfo
ERROR: Could not create esgf security database tables in esgcet

Sorry...
This action did not complete successfully
Please re-run this task until successful before continuing further
```

To investigate further...

**Comment by alaniwi** [comment]
Even with the conda environment activated, it is reading the wrong shared library.
```
28373 open("/usr/lib64/libpq.so.5", O_RDONLY) = 5
```
The activate script does not set `LD_LIBRARY_PATH`.

If I set it explicitly, I can do successfully:
```
LD_LIBRARY_PATH=/usr/local/conda/envs/esgf-pub/lib /usr/local/conda/bin/esgf_security_initialize --dburl dbsuper:the_password@localhost:5432/esgcet -c
```
(with `the_password` substituted)

but without it, get the error shown above.

**Comment by alaniwi** [comment]
There is a `_psycopg.so` in the root conda environment, which is linked to the system `libpq`:
```
# ldd /usr/local/conda/lib/python2.7/site-packages/psycopg2-2.7.3.1-py2.7-linux-x86_64.egg/psycopg2/_psycopg.so | grep libpq
        libpq.so.5 => /usr/lib64/libpq.so.5 (0x00007fd335505000)
```
This one appears to be causing the problem.

There is also a  `_psycopg.so` in the `esgf-pub` conda environment, which is correctly linked to the `libpq` in that environment:

```
# ldd /usr/local/conda/envs/esgf-pub/lib/python2.7/site-packages/psycopg2/_psycopg.so  | grep libpq
        libpq.so.5 => /usr/local/conda/envs/esgf-pub/lib/python2.7/site-packages/psycopg2/../../../libpq.so.5 (0x00007f9b8d26d000)
```
but this is not the one being used.

I am guessing that the problem is that forcing the path to `easy_install` to that in the root conda environment, in the line:
``` 
        $cdat_home/bin/easy_install ${esgf_security_egg_file}
```
is causing the problem. At present, `esgf_security_initialize` is being installed into `/usr/local/conda/bin` instead of `/usr/local/conda/envs/esgf-pub/bin`

I will try changing it to just pick up `easy_install` from `$PATH` (which, after activating the `esgf-pub` environment, will be `/usr/local/conda/envs/esgf-pub/bin/easy_install`.

**Comment by alaniwi** [comment]
Seems to be working. Installation still in progress, but this time no error reported, and it has set up the tables which it did not previously do.

**Comment by alaniwi** [comment]
I believe that this ticket can be closed when the above two-line pull request is merged. The installation was successful and the node is working properly, including the myproxy server which relies on the `esgf_security` database table (although I haven't explicitly tested the attribute service).

**Comment by SebastienDenvil** [comment]
Not strictly related to this issue but we would like to know if you tested a publish and download data protected by policies on the 2.5.14.

We have difficulties with this version here and would welcome your view.

**Comment by alaniwi** [comment]
@SebastienDenvil I will test it. So far, I only tested publish and download of the open access test dataset. Please create an issue for that, tag me on it, and I will respond there when I have tested it.

**Comment by sashakames** [comment]
I've pushed a fix to the devel branch of the mirror for 2.5.15 that seems to have corrected the problem.  lets ensure it works before closing this issue
